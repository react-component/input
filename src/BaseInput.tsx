import clsx from 'classnames';
import type { ReactElement, ReactNode } from 'react';
import React, { cloneElement, useRef } from 'react';
import type { BaseInputProps } from './interface';
import { hasAddon, hasPrefixSuffix } from './utils/commonUtils';

export interface HolderRef {
  /** Provider holder ref. Will return `null` if not wrap anything */
  nativeElement: HTMLElement | null;
}

const BaseInput = React.forwardRef<HolderRef, BaseInputProps>((props, ref) => {
  const {
    inputElement: inputEl,
    children,
    prefixCls,
    prefix,
    suffix,
    addonBefore,
    addonAfter,
    className,
    style,
    disabled,
    readOnly,
    focused,
    triggerFocus,
    allowClear,
    value,
    handleReset,
    hidden,
    classes,
    classNames,
    dataAttrs,
    styles,
    components,
    onClear,
  } = props;

  const inputElement = children ?? inputEl;

  const AffixWrapperComponent = components?.affixWrapper || 'span';
  const GroupWrapperComponent = components?.groupWrapper || 'span';
  const WrapperComponent = components?.wrapper || 'span';
  const GroupAddonComponent = components?.groupAddon || 'span';

  const containerRef = useRef<HTMLDivElement>(null);

  const onInputClick: React.MouseEventHandler = (e) => {
    if (containerRef.current?.contains(e.target as Element)) {
      triggerFocus?.();
    }
  };

  const hasAffix = hasPrefixSuffix(props);

  let element: ReactElement = cloneElement(inputElement, {
    value,
    className:
      clsx(inputElement.props.className, !hasAffix && classNames?.variant) ||
      null,
  });

  // ======================== Ref ======================== //
  const groupRef = useRef<HTMLDivElement>(null);

  React.useImperativeHandle(ref, () => ({
    nativeElement: groupRef.current || containerRef.current,
  }));

  // ================== Prefix & Suffix ================== //
  if (hasAffix) {
    // ================== Clear Icon ================== //
    let clearIcon: ReactNode = null;
    if (allowClear) {
      const needClear = !disabled && !readOnly && value;
      const clearIconCls = `${prefixCls}-clear-icon`;
      const iconNode =
        typeof allowClear === 'object' && allowClear?.clearIcon
          ? allowClear.clearIcon
          : 'âœ–';

      clearIcon = (
        <span
          onClick={(event) => {
            handleReset?.(event);
            onClear?.();
          }}
          // Do not trigger onBlur when clear input
          // https://github.com/ant-design/ant-design/issues/31200
          onMouseDown={(e) => e.preventDefault()}
          className={clsx(clearIconCls, {
            [`${clearIconCls}-hidden`]: !needClear,
            [`${clearIconCls}-has-suffix`]: !!suffix,
          })}
          role="button"
          tabIndex={-1}
        >
          {iconNode}
        </span>
      );
    }

    const affixWrapperPrefixCls = `${prefixCls}-affix-wrapper`;
    const affixWrapperCls = clsx(
      affixWrapperPrefixCls,
      {
        [`${prefixCls}-disabled`]: disabled,
        [`${affixWrapperPrefixCls}-disabled`]: disabled, // Not used, but keep it
        [`${affixWrapperPrefixCls}-focused`]: focused, // Not used, but keep it
        [`${affixWrapperPrefixCls}-readonly`]: readOnly,
        [`${affixWrapperPrefixCls}-input-with-clear-btn`]:
          suffix && allowClear && value,
      },
      classes?.affixWrapper,
      classNames?.affixWrapper,
      classNames?.variant,
    );

    const suffixNode = (suffix || allowClear) && (
      <span
        className={clsx(`${prefixCls}-suffix`, classNames?.suffix)}
        style={styles?.suffix}
      >
        {clearIcon}
        {suffix}
      </span>
    );

    element = (
      <AffixWrapperComponent
        className={affixWrapperCls}
        style={styles?.affixWrapper}
        onClick={onInputClick}
        {...dataAttrs?.affixWrapper}
        ref={containerRef}
      >
        {prefix && (
          <span
            className={clsx(`${prefixCls}-prefix`, classNames?.prefix)}
            style={styles?.prefix}
          >
            {prefix}
          </span>
        )}
        {element}
        {suffixNode}
      </AffixWrapperComponent>
    );
  }

  // ================== Addon ================== //
  if (hasAddon(props)) {
    const wrapperCls = `${prefixCls}-group`;
    const addonCls = `${wrapperCls}-addon`;
    const groupWrapperCls = `${wrapperCls}-wrapper`;

    const mergedWrapperClassName = clsx(
      `${prefixCls}-wrapper`,
      wrapperCls,
      classes?.wrapper,
      classNames?.wrapper,
    );

    const mergedGroupClassName = clsx(
      groupWrapperCls,
      {
        [`${groupWrapperCls}-disabled`]: disabled,
      },
      classes?.group,
      classNames?.groupWrapper,
    );

    // Need another wrapper for changing display:table to display:inline-block
    // and put style prop in wrapper
    element = (
      <GroupWrapperComponent className={mergedGroupClassName} ref={groupRef}>
        <WrapperComponent className={mergedWrapperClassName}>
          {addonBefore && (
            <GroupAddonComponent className={addonCls}>
              {addonBefore}
            </GroupAddonComponent>
          )}
          {element}
          {addonAfter && (
            <GroupAddonComponent className={addonCls}>
              {addonAfter}
            </GroupAddonComponent>
          )}
        </WrapperComponent>
      </GroupWrapperComponent>
    );
  }

  // `className` and `style` are always on the root element
  return React.cloneElement(element, {
    className: clsx(element.props?.className, className) || null,
    style: {
      ...element.props?.style,
      ...style,
    },
    hidden,
  });
});

export default BaseInput;
