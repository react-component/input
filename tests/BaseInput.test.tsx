import { fireEvent, render } from '@testing-library/react';
import type { ChangeEvent, FC } from 'react';
import React, { useState } from 'react';
import BaseInput, { type HolderRef } from '../src/BaseInput';

describe('BaseInput', () => {
  it('should render perfectly', () => {
    const { container } = render(
      <BaseInput prefixCls="rc-input">
        <input />
      </BaseInput>,
    );
    expect(container).toMatchSnapshot();
  });

  it('prefix and suffix should render properly', () => {
    const { container } = render(
      <div>
        <BaseInput prefixCls="rc-input" prefix="prefix">
          <input />
        </BaseInput>
        <br />
        <br />
        <BaseInput prefixCls="rc-input" suffix="suffix">
          <input />
        </BaseInput>
      </div>,
    );
    expect(container).toMatchSnapshot();
  });

  it('addon should render properly', () => {
    const { container } = render(
      <div>
        <BaseInput prefixCls="rc-input" addonBefore="addonBefore">
          <input />
        </BaseInput>
        <br />
        <br />
        <BaseInput prefixCls="rc-input" addonAfter="addonAfter">
          <input />
        </BaseInput>
      </div>,
    );
    expect(container).toMatchSnapshot();
  });

  it('allowClear should work', () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const onFocus = jest.fn();

    const Demo: FC = () => {
      const [value, setValue] = useState<string>('');

      const handleReset = () => {
        setValue('');
      };

      const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
        onChange();
        setValue(e.target.value);
      };

      return (
        <BaseInput
          prefixCls="rc-input"
          allowClear={{ clearIcon: '✖' }}
          value={value}
          handleReset={handleReset}
        >
          <input onChange={handleChange} onBlur={onBlur} onFocus={onFocus} />
        </BaseInput>
      );
    };

    const { container } = render(<Demo />);

    const inputEl = container.querySelector('input');
    fireEvent.focus(inputEl!);
    expect(onFocus).toHaveBeenCalledTimes(1);

    fireEvent.change(inputEl!, { target: { value: 'some text' } });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(inputEl!.value).toBe('some text');

    const clearIcon = container.querySelector('.rc-input-clear-icon');
    fireEvent.mouseDown(clearIcon!);
    fireEvent.click(clearIcon!);
    fireEvent.mouseUp(clearIcon!);
    expect(onBlur).not.toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(inputEl!.value).toBe('');
  });

  it('should display clearIcon correctly', () => {
    const { container, rerender } = render(
      <BaseInput prefixCls="rc-input" allowClear>
        <input />
      </BaseInput>,
    );
    const clearIcon = container.querySelector('.rc-input-clear-icon');
    expect(clearIcon?.innerHTML).toBe('✖');
    rerender(
      <BaseInput prefixCls="rc-input" allowClear={{ clearIcon: 'clear' }}>
        <input />
      </BaseInput>,
    );
    expect(clearIcon?.innerHTML).toBe('clear');
  });

  it('should focus input when click prefix', () => {
    const inputRef = React.createRef<HTMLInputElement>();
    const { container } = render(
      <BaseInput
        prefixCls="rc-input"
        triggerFocus={() => inputRef.current?.focus()}
        prefix="$"
      >
        <input ref={inputRef} />
      </BaseInput>,
    );
    fireEvent.click(container.querySelector('.rc-input-affix-wrapper')!);
    expect(document.activeElement).toBe(container.querySelector('input'));
  });

  it('should support data-*', () => {
    const { container } = render(
      <BaseInput
        prefixCls="rc-input"
        prefix="prefix"
        dataAttrs={{
          affixWrapper: {
            'data-test': 'test',
          },
        }}
      >
        <input />
      </BaseInput>,
    );
    expect(
      container
        .querySelector('.rc-input-affix-wrapper')
        ?.getAttribute('data-test'),
    ).toBe('test');
  });

  it('should apply className to inputElement', () => {
    const { container } = render(
      <BaseInput prefixCls="rc-input" className="test-base">
        <input className="test" />
      </BaseInput>,
    );
    expect(container.querySelector('.test-base')).toBeTruthy();
    expect(container.querySelector('.test')).toBeTruthy();
  });

  it('should not pass className to inputElement when has addon', () => {
    const { container } = render(
      <BaseInput prefixCls="rc-input" className="test-base" addonBefore="addon">
        <input className="test" />
      </BaseInput>,
    );
    expect(
      container.querySelector('input')?.classList.contains('test'),
    ).toBeTruthy();
    expect(
      container.querySelector('input')?.classList.contains('test-base'),
    ).toBeFalsy();
  });

  it('should correct render with prefix and addon', () => {
    const { container } = render(
      <BaseInput prefixCls="rc-input" prefix="prefix" addonBefore="addon">
        <input />
      </BaseInput>,
    );
    expect(container).toMatchSnapshot();
  });

  it('support use div as basic element', () => {
    const { container } = render(
      <BaseInput
        prefixCls="rc-input"
        prefix="prefix"
        addonBefore="addon"
        components={{
          affixWrapper: 'div',
          groupWrapper: 'div',
          wrapper: 'div',
          groupAddon: 'div',
        }}
      >
        <input />
      </BaseInput>,
    );
    expect(container).toMatchSnapshot();
  });

  it('styles should work', () => {
    const container = render(
      <BaseInput
        prefixCls="rc-input"
        prefix="prefix"
        addonBefore="addon"
        styles={{
          affixWrapper: {
            color: 'red',
          },
          prefix: {
            color: 'blue',
          },
        }}
      >
        <input />
      </BaseInput>,
    );

    expect(
      container.container.querySelector<HTMLSpanElement>(
        '.rc-input-affix-wrapper',
      )?.style.color,
    ).toBe('red');
    expect(
      container.container.querySelector<HTMLSpanElement>('.rc-input-prefix')
        ?.style.color,
    ).toBe('blue');
  });

  it('with addon and disabled', () => {
    const { container } = render(
      <BaseInput prefixCls="rc-input" addonBefore="addon" disabled>
        <input />
      </BaseInput>,
    );

    expect(container.firstChild).toHaveClass('rc-input-group-wrapper-disabled');
  });

  it('variant cls', () => {
    const { container, rerender } = render(
      <BaseInput
        prefixCls="rc-input"
        prefix="$"
        classNames={{ variant: 'test-variant' }}
        disabled
      >
        <input />
      </BaseInput>,
    );

    expect(container.querySelector('.rc-input-affix-wrapper')).toHaveClass(
      'test-variant',
    );
    expect(container.querySelector('input')).not.toHaveClass('test-variant');

    rerender(
      <BaseInput
        prefixCls="rc-input"
        classNames={{ variant: 'test-variant' }}
        disabled
      >
        <input />
      </BaseInput>,
    );

    expect(container.querySelector('.rc-input-affix-wrapper')).toBeFalsy();
    expect(container.querySelector('input')).toHaveClass('test-variant');
  });

  describe('ref', () => {
    it('prefix', () => {
      const holderRef = React.createRef<HolderRef>();
      const { container } = render(
        <BaseInput prefixCls="rc-input" prefix="prefix" ref={holderRef}>
          <input />
        </BaseInput>,
      );
      expect(holderRef.current?.nativeElement).toBe(
        container.querySelector('.rc-input-affix-wrapper'),
      );
    });

    it('addon', () => {
      const holderRef = React.createRef<HolderRef>();
      const { container } = render(
        <BaseInput prefixCls="rc-input" addonAfter="after" ref={holderRef}>
          <input />
        </BaseInput>,
      );

      expect(holderRef.current?.nativeElement).toBe(
        container.querySelector('.rc-input-group-wrapper'),
      );
    });

    it('mix', () => {
      const holderRef = React.createRef<HolderRef>();
      const { container } = render(
        <BaseInput
          prefixCls="rc-input"
          suffix="suffix"
          addonAfter="after"
          ref={holderRef}
        >
          <input />
        </BaseInput>,
      );

      expect(holderRef.current?.nativeElement).toBe(
        container.querySelector('.rc-input-group-wrapper'),
      );
    });

    it('support onClear', () => {
      const onClear = jest.fn();
      const { container } = render(
        <BaseInput prefixCls="rc-input" onClear={onClear} allowClear>
          <input defaultValue="test" />
        </BaseInput>,
      );
      fireEvent.click(
        container.querySelector<HTMLSpanElement>('.rc-input-clear-icon')!,
      );
      expect(onClear).toHaveBeenCalled();
    });
  });
});
