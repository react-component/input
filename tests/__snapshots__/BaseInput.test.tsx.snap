// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseInput addon should render properly 1`] = `
<div>
  <div>
    <span
      class="rc-input-group-wrapper"
    >
      <span
        class="rc-input-wrapper rc-input-group"
      >
        <span
          class="rc-input-group-addon"
        >
          addonBefore
        </span>
        <input
          value=""
        />
      </span>
    </span>
    <br />
    <br />
    <span
      class="rc-input-group-wrapper"
    >
      <span
        class="rc-input-wrapper rc-input-group"
      >
        <input
          value=""
        />
        <span
          class="rc-input-group-addon"
        >
          addonAfter
        </span>
      </span>
    </span>
  </div>
</div>
`;

exports[`BaseInput prefix and suffix should render properly 1`] = `
<div>
  <div>
    <span
      class="rc-input-affix-wrapper"
    >
      <span
        class="rc-input-prefix"
      >
        prefix
      </span>
      <input
        value=""
      />
    </span>
    <br />
    <br />
    <span
      class="rc-input-affix-wrapper"
    >
      <input
        value=""
      />
      <span
        class="rc-input-suffix"
      >
        suffix
      </span>
    </span>
  </div>
</div>
`;

exports[`BaseInput should correct render with prefix and addon 1`] = `
<div>
  <span
    class="rc-input-group-wrapper"
  >
    <span
      class="rc-input-wrapper rc-input-group"
    >
      <span
        class="rc-input-group-addon"
      >
        addon
      </span>
      <span
        class="rc-input-affix-wrapper"
      >
        <span
          class="rc-input-prefix"
        >
          prefix
        </span>
        <input
          value=""
        />
      </span>
    </span>
  </span>
</div>
`;

exports[`BaseInput should render perfectly 1`] = `
<div>
  <input
    value=""
  />
</div>
`;

exports[`BaseInput support use div as basic element 1`] = `
<div>
  <div
    class="rc-input-group-wrapper"
  >
    <div
      class="rc-input-wrapper rc-input-group"
    >
      <div
        class="rc-input-group-addon"
      >
        addon
      </div>
      <div
        class="rc-input-affix-wrapper"
      >
        <span
          class="rc-input-prefix"
        >
          prefix
        </span>
        <input
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
